starvine/vine/C_vine.py

```diff
@@ -108,6 +109,141 @@ def buildDeepTrees(self, level=1):
+                sum_pdf += \
+                    self.vine[lvl].tree.adj[nonRootID][rootID]["pc"].copulaModel.pdf( \
+                    u, v) / tree_wgt
+        return sum_pdf
+        """
+        ln_pdf = np.zeros(len(x))
+        total_n_copula = np.math.factorial(self.nLevels)
+        for u_id, v_id in self.vine[0].tree.edges(data=False):
+            x_lvl = x
+
+            rootID = self.vine[0].rootNodeID
+            if u_id is not rootID:
+                nonRootID = u_id
+            else:
+                nonRootID = v_id
+
+            # u, v = np.asarray(x_lvl[u_id]), np.asarray(x_lvl[v_id])
+            u, v = np.asarray(x_lvl[nonRootID]), np.asarray(x_lvl[rootID])
+
+            ln_pdf += \
+                np.log(self.vine[0].tree.adj[nonRootID][rootID]["pc"].copulaModel.pdf( \
+                u, v))
+
+        for lvl in range(1, self.nLevels):
+            for u_id, v_id in self.vine[lvl].tree.edges(data=False):
+                x_lvl = vine_cond_x[lvl - 1]
+
+                rootID = self.vine[lvl].rootNodeID
+                if u_id is not rootID:
+                    nonRootID = u_id
+                else:
+                    nonRootID = v_id
+
+                # u, v = np.asarray(x_lvl[u_id]), np.asarray(x_lvl[v_id])
+                u, v = np.asarray(x_lvl[nonRootID]), np.asarray(x_lvl[rootID])
+
+                ln_pdf += \
+                    np.log(self.vine[lvl].tree.adj[nonRootID][rootID]["pc"].copulaModel.pdf( \
+                    u, v))
+        return ln_pdf
+
+    def _pdf(self, x, **kwargs):
+        assert isinstance(x, DataFrame)
+        return np.exp(self._lnpdf(x, **kwargs))
+
+    def _cdf(self, x, **kwargs):
+        import quadpy
+        assert isinstance(x, DataFrame)
+        cdf_vector = np.zeros(len(x))
+        points_vector = np.ones(len(x)) * 0.0
+        def _pdf(*x):
+            # x should always have even len
+            col_labels = x[int(len(x)/2):]
+            x_dframe = DataFrame()
+            for i, col_label in enumerate(col_labels):
+                x_dframe[col_label] = [x[i]]
+            pdf_x = self._pdf(x_dframe)
+            print(x[0:int(len(x)/2)], pdf_x)
+            return pdf_x
+
+        def f_pdf(x_np):
+            nps = x_np.shape
+            x_dframe = DataFrame()
+            for k in range(nps[0]):
+                x_dframe[x.columns[k]] = x_np[k, :]
+            pdf_x = self._pdf(x_dframe)
+            pdf_x[np.isnan(pdf_x)] = np.max(pdf_x)
+            return pdf_x
+
+        edge_tol = 1e-2
+        for i, x_i in x.iterrows():
+            ranges = np.zeros((len(x_i), 2)) + edge_tol
+            col_labels = []
+            j=0
+            for col_name, col_data  in x_i.iteritems():
+                col_labels.append(col_name)
+                ranges[j, 1] = np.clip(col_data, edge_tol, 1. - 0.1 * edge_tol)
+                j += 1
+
+            dim = len(x.columns)
+            scheme = quadpy.ncube.stroud_cn_5_9(dim)
```

Comment:
probably best to rework this.  Split hypercube up into chunks, then integrate pdf over each chunk.  Try scipy.spatial.Delaunay to chop up integration domain, then use quadpy to perform the integration on each chunk.